pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1183,53 +1198,166 @@ func (c *Reconciler) updatePipelineRunStatusFromInformer(ctx context.Context, pr
+		// Verify that we don't populate TaskRun statuses for "minimal"
+		if len(prs.TaskRuns) > 0 {
+			err = multierror.Append(err, fmt.Errorf("expected no TaskRun statuses with embedded-status=minimal, but found %d", len(prs.TaskRuns)))
+		}
+		// Verify that we don't populate Run statuses for "minimal"
+		if len(prs.Runs) > 0 {
+			err = multierror.Append(err, fmt.Errorf("expected no Run statuses with embedded-status=minimal, but found %d", len(prs.Runs)))
+		}
+		for _, cr := range prs.ChildReferences {
+			switch cr.Kind {
+			case "TaskRun", "Run":
+				continue
+			default:
+				err = multierror.Append(err, fmt.Errorf("child with name %s has unknown kind %s", cr.Name, cr.Kind))
+			}
 		}
-	} else {
-		pr.Status.TaskRuns = make(map[string]*v1beta1.PipelineRunTaskRunStatus)
 	}
-	// Loop over all the TaskRuns associated to Tasks
-	for _, taskrun := range trs {
+
+	// Verify that the TaskRun and Run statuses match the ChildReferences for "both"
+	if cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus {
+		if len(prs.ChildReferences) != len(prs.TaskRuns)+len(prs.Runs) {
+			err = multierror.Append(err, fmt.Errorf("expected the same number of ChildReferences as the number of combined TaskRun and Run statuses (%d), but found %d",
+				len(prs.TaskRuns)+len(prs.Runs),
+				len(prs.ChildReferences)))
+		}
+
+		for _, cr := range prs.ChildReferences {
+			switch cr.Kind {
+			case "TaskRun":
+				tr, ok := prs.TaskRuns[cr.Name]
+				if !ok {
+					err = multierror.Append(err, fmt.Errorf("embedded-status is 'both', and child TaskRun with name %s found in ChildReferences only", cr.Name))
+				} else if cr.PipelineTaskName != tr.PipelineTaskName {
+					err = multierror.Append(err,
+						fmt.Errorf("child TaskRun with name %s has PipelineTask name %s in ChildReferences and %s in TaskRuns",
+							cr.Name, cr.PipelineTaskName, tr.PipelineTaskName))
+				}
+			case "Run":
+				r, ok := prs.Runs[cr.Name]
+				if !ok {
+					err = multierror.Append(err, fmt.Errorf("embedded-status is 'both', and child Run with name %s found in ChildReferences only", cr.Name))
+				} else if cr.PipelineTaskName != r.PipelineTaskName {
+					err = multierror.Append(err,
+						fmt.Errorf("child Run with name %s has PipelineTask name %s in ChildReferences and %s in Runs",
+							cr.Name, cr.PipelineTaskName, r.PipelineTaskName))
+				}
+			default:
+				err = multierror.Append(err, fmt.Errorf("child with name %s has unknown kind %s", cr.Name, cr.Kind))
+			}
+		}
+	}
+
+	return err
+}
+
+// filterTaskRunsForPipelineRun goes through the list of TaskRuns, skipping any which aren't owned by the given PipelineRun,
+// and then splitting the remainder into a slice of TaskRuns directly specified in the PipelineRun and a map of condition
+// TaskRuns keyed by the PipelineTask name they relate to.
+func filterTaskRunsForPipelineRun(logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, trs []*v1beta1.TaskRun) ([]*v1beta1.TaskRun, map[string][]*v1beta1.TaskRun) {
```

Comment:
could you please reword this comment to describe the inputs and outputs of the function rather than its implementation?

Reply:
👍